## Load Data

In [1]:
#Data :https://www.eea.europa.eu/data-and-maps/figures/correlation-between-fertility-and-female-education
import pandas as pd
import numpy as np
import ssl
ssl._create_default_https_context = ssl._create_unverified_context

fem_lit = pd.read_csv('https://assets.datacamp.com/production/course_1392/datasets/literacy_birth_rate.csv')
df = pd.read_excel('./slides/TREND01-5G-educ-fertility-bubbles.xls',sheet_name='data COMPILATION',header=7)
df.head()

Country  Continent female literacy fertility    population
0      Chine       ASI            90.5     1.769  1.324655e+09
1       Inde       ASI            50.8     2.682  1.139965e+09
2        USA       NAM              99     2.077  3.040600e+08
3  Indonésie       ASI            88.8     2.132  2.273451e+08
4     Brésil       LAT            90.2     1.827  1.919715e+08

In [3]:
appl = pd.read_csv('https://assets.datacamp.com/production/course_1392/datasets/aapl.csv')
appl.head()

Unnamed: 0  adj_close   close        date    high     low    open    volume
0           0      31.68  130.31  2000-03-01  132.06  118.50  118.56  38478000
1           1      29.66  122.00  2000-03-02  127.94  120.69  127.00  11136800
2           2      31.12  128.00  2000-03-03  128.23  120.00  124.87  11565200
3           3      30.56  125.69  2000-03-06  129.13  125.00  126.00   7520000
4           4      29.87  122.87  2000-03-07  127.44  121.12  126.44   9767600

In [4]:
auto = pd.read_csv('https://assets.datacamp.com/production/course_1392/datasets/auto-mpg.csv')
auto.head()

mpg  cyl  displ   hp  weight  accel  yr  origin              name  color  \
0  18.0    6  250.0   88    3139   14.5  71      US      ford mustang   blue   
1   9.0    8  304.0  193    4732   18.5  70      US          hi 1200d   blue   
2  36.1    4   91.0   60    1800   16.4  78    Asia  honda civic cvcc    red   
3  18.5    6  250.0   98    3525   19.0  77      US      ford granada   blue   
4  34.3    4   97.0   78    2188   15.8  80  Europe         audi 4000  green   

   size  
0  15.0  
1  20.0  
2  10.0  
3  15.0  
4  10.0

In [5]:
gapminder = pd.read_csv('https://assets.datacamp.com/production/course_1392/datasets/gapminder_tidy.csv')
gapminder.head()

Country  Year  fertility    life  population  child_mortality     gdp  \
0  Afghanistan  1964      7.671  33.639  10474903.0            339.7  1182.0   
1  Afghanistan  1965      7.671  34.152  10697983.0            334.1  1182.0   
2  Afghanistan  1966      7.671  34.662  10927724.0            328.7  1168.0   
3  Afghanistan  1967      7.671  35.170  11163656.0            323.3  1173.0   
4  Afghanistan  1968      7.671  35.674  11411022.0            318.1  1187.0   

       region  
0  South Asia  
1  South Asia  
2  South Asia  
3  South Asia  
4  South Asia

In [7]:
glucose = pd.read_csv('https://assets.datacamp.com/production/course_1392/datasets/glucose.csv')
glucose.head()

datetime   isig  glucose
0  2010-10-07 00:03:00  22.10      150
1  2010-10-07 00:08:00  21.46      152
2  2010-10-07 00:13:00  21.06      149
3  2010-10-07 00:18:00  20.96      147
4  2010-10-07 00:23:00  21.52      148

In [8]:
olympic = pd.read_csv('https://assets.datacamp.com/production/course_1392/datasets/sprint.csv')
olympic.head()

Name Country   Medal  Time  Year        color
0        Usain Bolt     JAM    GOLD  9.63  2012    goldenrod
1       Yohan Blake     JAM  SILVER  9.75  2012       silver
2     Justin Gatlin     USA  BRONZE  9.79  2012  saddlebrown
3        Usain Bolt     JAM    GOLD  9.69  2008    goldenrod
4  Richard Thompson     TRI  SILVER  9.89  2008       silver

## Chapter 1: Basic plotting with Bokeh

An introduction to basic plotting with Bokeh. You will create your first plots, learn about different data formats Bokeh understands, and make visual customizations for selections and mouse hovering.
### Plotting with Glyphs
#### A simple scatter plot
In this example, you're going to make a scatter plot of female literacy vs fertility using data from the European Environmental Agency. This dataset highlights that countries with low female literacy have high birthrates. The x-axis data has been loaded for you as fertility and the y-axis data has been loaded as female_literacy.

Your job is to create a figure, assign x-axis and y-axis labels, and plot female_literacy vs fertility using the circle glyph.

After you have created the figure, in this exercise and the ones to follow, play around with it! Explore the different options available to you on the tab to the right, such as "Pan", "Box Zoom", and "Wheel Zoom". You can click on the question mark sign for more details on any of these tools.

Note: You may have to scroll down to view the lower portion of the figure.

In [47]:
fertility = df['fertility'].values
female_literacy = df['female literacy'].values
print(df['Continent'].unique())

['ASI' 'NAM' 'LAT' 'AF' 'EUR' 'OCE' nan 'Continent' 'WORLD']


In [10]:
# Import figure from bokeh.plotting
from bokeh.plotting import figure
# Import output_file and show from bokeh.io
from bokeh.io import output_file, show,output_notebook
output_notebook()

Loading BokehJS ...

In [11]:
# Create the figure: p
p = figure(x_axis_label='fertility (children per woman)', y_axis_label='female_literacy (% population)')

# Add a circle glyph to the figure p
p.circle(fertility,female_literacy)

# Call the output_file() function and specify the name of the file
output_file('./output/fert_lit.html')

# Display the plot
show(p)

#### A scatter plot with different shapes
By calling multiple glyph functions on the same figure object, we can overlay multiple data sets in the same figure.

In this exercise, you will plot female literacy vs fertility for two different regions, Africa and Latin America. Each set of x and y data has been loaded separately for you as fertility_africa, female_literacy_africa, fertility_latinamerica, and female_literacy_latinamerica.

Your job is to plot the Latin America data with the circle() glyph, and the Africa data with the x() glyph.

figure has already been imported for you from bokeh.plotting.

In [48]:
# Create the figure: p
p = figure(x_axis_label='fertility', y_axis_label='female_literacy (% population)')

fertility_latinamerica = df[df['Continent'] == 'LAT']['fertility'].values
female_literacy_latinamerica = df[df['Continent'] == 'LAT']['female literacy'].values
fertility_africa = df[df['Continent'] == 'AF']['fertility'].values
female_literacy_africa = df[df['Continent'] == 'AF']['female literacy'].values


# Add a circle glyph to the figure p
p.circle(fertility_latinamerica,female_literacy_latinamerica)

# Add an x glyph to the figure p
p.x(fertility_africa,female_literacy_africa)

# Specify the name of the file
output_file('./output/fert_lit_separate.html')

# Display the plot
show(p)

#### Customizing your scatter plots
The three most important arguments to customize scatter glyphs are color, size, and alpha. Bokeh accepts colors as hexadecimal strings, tuples of RGB values between 0 and 255, and any of the 147 CSS color names. Size values are supplied in screen space units with 100 meaning the size of the entire figure.

The alpha parameter controls transparency. It takes in floating point numbers between 0.0, meaning completely transparent, and 1.0, meaning completely opaque.

In this exercise, you'll plot female literacy vs fertility for Africa and Latin America as red and blue circle glyphs, respectively.

In [49]:
# Create the figure: p
p = figure(x_axis_label='fertility (children per woman)', y_axis_label='female_literacy (% population)')

# Add a blue circle glyph to the figure p
p.circle(fertility_latinamerica, female_literacy_latinamerica, color='blue', alpha=.8, size=10)

# Add a red circle glyph to the figure p
p.circle(fertility_africa, female_literacy_africa, color='red', alpha=.8, size=10)

# Specify the name of the file
output_file('./output/fert_lit_separate_colors.html')

# Display the plot
show(p)


### Additional Glyphs
#### Lines
We can draw lines on Bokeh plots with the line() glyph function.

In this exercise, you'll plot the daily adjusted closing price of Apple Inc.'s stock (AAPL) from 2000 to 2013.

The data points are provided for you as lists. date is a list of datetime objects to plot on the x-axis and price is a list of prices to plot on the y-axis.

Since we are plotting dates on the x-axis, you must add x_axis_type='datetime' when creating the figure object.

In [14]:
appl.date = appl.date.astype('datetime64[ns]')
date = appl.date.values
price = appl.close.values
date[0:10],price[0:10]

(array(['2000-03-01T00:00:00.000000000', '2000-03-02T00:00:00.000000000',
        '2000-03-03T00:00:00.000000000', '2000-03-06T00:00:00.000000000',
        '2000-03-07T00:00:00.000000000', '2000-03-08T00:00:00.000000000',
        '2000-03-09T00:00:00.000000000', '2000-03-10T00:00:00.000000000',
        '2000-03-13T00:00:00.000000000', '2000-03-14T00:00:00.000000000'],
       dtype='datetime64[ns]'),
 array([130.31, 122.  , 128.  , 125.69, 122.87, 122.  , 122.25, 125.75,
        121.31, 114.25]))

In [15]:
# Create a figure with x_axis_type="datetime": p
p = figure(x_axis_type="datetime", x_axis_label='Date', y_axis_label='US Dollars')

# Plot date along the x axis and price along the y axis
p.line(date,price)

# Specify the name of the output file and show the result
output_file('./output/line.html')
show(p)

#### Lines and markers
Lines and markers can be combined by plotting them separately using the same data points.

In this exercise, you'll plot a line and circle glyph for the AAPL stock prices. Further, you'll adjust the fill_color keyword argument of the circle() glyph function while leaving the line_color at the default value.

The date and price lists are provided. The Bokeh figure object p that you created in the previous exercise has also been provided.

In [16]:
date = date[-100:]
price = price[-100:]
# Create a figure with x_axis_type='datetime': p
p = figure(x_axis_type='datetime', x_axis_label='Date', y_axis_label='US Dollars')

# Plot date along the x-axis and price along the y-axis
p.line(date,price)

# With date on the x-axis and price on the y-axis, add a white circle glyph of size 4
p.circle(date, price, fill_color='white', size=4)

# Specify the name of the output file and show the result
output_file('./output/line2.html')
show(p)

#### Patches
In Bokeh, extended geometrical shapes can be plotted by using the patches() glyph function. The patches glyph takes as input a list-of-lists collection of numeric values specifying the vertices in x and y directions of each distinct patch to plot.

In this exercise, you will plot the state borders of Arizona, Colorado, New Mexico and Utah. The latitude and longitude vertices for each state have been prepared as lists.

Your job is to plot longitude on the x-axis and latitude on the y-axis. The figure object has been created for you as p.

##### Longitude data

In [17]:
az_lons = [-114.63332,
 -114.63349,
 -114.63423,
 -114.60899,
 -114.63064,
 -114.57354,
 -114.58031,
 -114.61121,
 -114.6768,
 -114.66076,
 -114.65449,
 -114.68702,
 -114.69704,
 -114.70415,
 -114.67489,
 -114.70883,
 -114.74365,
 -114.73513,
 -114.6729,
 -114.51122,
 -114.32346,
 -114.22646,
 -114.1139,
 -114.04404,
 -114.04338,
 -114.04736,
 -114.05014,
 -114.0506,
 -114.0506,
 -114.05052,
 -113.94557,
 -113.86852,
 -113.62465,
 -113.4727,
 -113.32097,
 -113.17698,
 -113.02079,
 -112.99281,
 -112.96895,
 -112.75086,
 -112.48455,
 -112.32985,
 -111.99142,
 -111.58602,
 -111.39598,
 -111.2523,
 -111.03957,
 -110.73783,
 -110.54945,
 -110.272,
 -110.13851,
 -109.83491,
 -109.43568,
 -109.26993,
 -109.04538,
 -109.04522,
 -109.04522,
 -109.04531,
 -109.04544,
 -109.04547,
 -109.04579,
 -109.04575,
 -109.04601,
 -109.04578,
 -109.04606,
 -109.04621,
 -109.04636,
 -109.04662,
 -109.04644,
 -109.04598,
 -109.04603,
 -109.04633,
 -109.04692,
 -109.047,
 -109.04691,
 -109.0474,
 -109.04762,
 -109.04764,
 -109.04811,
 -109.04905,
 -109.04911,
 -109.05004,
 -109.0587,
 -109.25062,
 -109.30069,
 -109.33682,
 -109.38186,
 -109.45105,
 -109.5287,
 -109.62562,
 -109.79302,
 -109.97582,
 -110.20503,
 -110.49327,
 -110.56918,
 -110.65415,
 -110.77828,
 -110.87564,
 -110.93778,
 -110.94286,
 -110.97553,
 -111.12565,
 -111.24082,
 -111.29191,
 -111.32558,
 -111.3574,
 -111.38483,
 -111.44337,
 -111.47861,
 -111.49725,
 -111.53479,
 -111.56975,
 -111.62412,
 -111.66,
 -111.73365,
 -111.79498,
 -111.9182,
 -111.97172,
 -111.99115,
 -112.02937,
 -112.09379,
 -112.13972,
 -112.15906,
 -112.21295,
 -112.32605,
 -112.39932,
 -112.43603,
 -112.52208,
 -112.57141,
 -112.63294,
 -112.67695,
 -112.72356,
 -112.75567,
 -112.8055,
 -112.83423,
 -112.8711,
 -112.90863,
 -113.20884,
 -113.2279,
 -113.30314,
 -113.61086,
 -113.78489,
 -113.90756,
 -113.97121,
 -114.11135,
 -114.20719,
 -114.25559,
 -114.28755,
 -114.38472,
 -114.61337,
 -114.77804,
 -114.81394,
 -114.81518,
 -114.80524,
 -114.81037,
 -114.81335,
 -114.80551,
 -114.80529,
 -114.79282,
 -114.79206,
 -114.79555,
 -114.81362,
 -114.80894,
 -114.80404,
 -114.80093,
 -114.80804,
 -114.80891,
 -114.80192,
 -114.79518,
 -114.7819,
 -114.77309,
 -114.76427,
 -114.74805,
 -114.74638,
 -114.74505,
 -114.74456,
 -114.74203,
 -114.74,
 -114.73874,
 -114.73062,
 -114.72924,
 -114.72377,
 -114.71994,
 -114.71919,
 -114.69096,
 -114.63501,
 -114.58576,
 -114.46563,
 -114.48131,
 -114.62973,
 -114.68157,
 -114.72123,
 -114.61185,
 -114.5402,
 -114.49649,
 -114.52801,
 -114.51318,
 -114.49813,
 -114.4355,
 -114.35765,
 -114.26017,
 -114.14737,
 -114.29195,
 -114.38169,
 -114.44166,
 -114.48236,
 -114.56953,
 -114.63305]

co_lons=[-109.04984,
 -109.06017,
 -109.06015,
 -109.05655,
 -109.05305,
 -109.05158,
 -109.05119,
 -109.05077,
 -109.05132,
 -109.05077,
 -109.05087,
 -109.05088,
 -109.05093,
 -109.05088,
 -109.05051,
 -109.04899,
 -109.04907,
 -109.05008,
 -109.03134,
 -108.83854,
 -108.6996,
 -108.59802,
 -108.46524,
 -108.25764,
 -108.10567,
 -107.91411,
 -107.75063,
 -107.55479,
 -107.35937,
 -107.27483,
 -107.12561,
 -106.9166,
 -106.59389,
 -106.32621,
 -106.06118,
 -105.82273,
 -105.60473,
 -105.46928,
 -105.2756,
 -105.07514,
 -104.86787,
 -104.58704,
 -104.24506,
 -104.0535,
 -104.05325,
 -104.05153,
 -103.90732,
 -103.61529,
 -103.46471,
 -103.30624,
 -103.00202,
 -102.88728,
 -102.76668,
 -102.55879,
 -102.38345,
 -102.23951,
 -102.12861,
 -102.05156,
 -102.05154,
 -102.05152,
 -102.05145,
 -102.05143,
 -102.05144,
 -102.05144,
 -102.0515,
 -102.0515,
 -102.05152,
 -102.05131,
 -102.05145,
 -102.05176,
 -102.05176,
 -102.05174,
 -102.05174,
 -102.05139,
 -102.0506,
 -102.04955,
 -102.04907,
 -102.04875,
 -102.04802,
 -102.04541,
 -102.04514,
 -102.04524,
 -102.04494,
 -102.04501,
 -102.04479,
 -102.04452,
 -102.04447,
 -102.04431,
 -102.04456,
 -102.04402,
 -102.04286,
 -102.04166,
 -102.04179,
 -102.04178,
 -102.04192,
 -102.04196,
 -102.04199,
 -102.0421,
 -102.04209,
 -102.04538,
 -102.05418,
 -102.07425,
 -102.09059,
 -102.17511,
 -102.19751,
 -102.24848,
 -102.26493,
 -102.29933,
 -102.34378,
 -102.35537,
 -102.40232,
 -102.49444,
 -102.52468,
 -102.57091,
 -102.62547,
 -102.68337,
 -102.7059,
 -102.77324,
 -102.81507,
 -102.86545,
 -102.89315,
 -102.9279,
 -102.97961,
 -102.98698,
 -103.00214,
 -103.0022,
 -103.01356,
 -103.26061,
 -103.53892,
 -103.92627,
 -104.17265,
 -104.49245,
 -104.76311,
 -105.15658,
 -105.50752,
 -105.6576,
 -105.93761,
 -106.15386,
 -106.4654,
 -106.61906,
 -106.73137,
 -106.89142,
 -107.00562,
 -107.25094,
 -107.4121,
 -107.49519,
 -107.8669,
 -108.20233,
 -108.52833,
 -108.7496,
 -108.8978,
 -109.04518,
 -109.04522,
 -109.04531,
 -109.04519,
 -109.04583,
 -109.04582,
 -109.04191,
 -109.04159,
 -109.04304,
 -109.04232,
 -109.0498]
nm_lons = [-103.55583,
 -104.00265,
 -104.64165,
 -105.14679,
 -105.90075,
 -106.55721,
 -106.63119,
 -106.62216,
 -106.63325,
 -106.61103,
 -106.54568,
 -106.52834,
 -106.52861,
 -106.53181,
 -106.55963,
 -106.56993,
 -106.60042,
 -106.61408,
 -106.62967,
 -106.67613,
 -106.68194,
 -106.7505,
 -106.75874,
 -106.77057,
 -106.82052,
 -106.87555,
 -106.89949,
 -106.93928,
 -106.96546,
 -106.98754,
 -106.99317,
 -106.99823,
 -106.99882,
 -107.00056,
 -107.12556,
 -107.2773,
 -107.29766,
 -107.3093,
 -107.32402,
 -107.35469,
 -107.37629,
 -107.39716,
 -107.42244,
 -107.47074,
 -107.50261,
 -107.53069,
 -107.5835,
 -107.62497,
 -107.6266,
 -107.62774,
 -107.62988,
 -107.63241,
 -107.63786,
 -107.64356,
 -107.65173,
 -107.66987,
 -107.70084,
 -107.70821,
 -107.75058,
 -107.7889,
 -107.84105,
 -107.86238,
 -107.8711,
 -107.88551,
 -107.90473,
 -107.964,
 -108.00059,
 -108.044,
 -108.05664,
 -108.07876,
 -108.08267,
 -108.1052,
 -108.15227,
 -108.17096,
 -108.1986,
 -108.20839,
 -108.20844,
 -108.20841,
 -108.2084,
 -108.20838,
 -108.20837,
 -108.2083,
 -108.20814,
 -108.20819,
 -108.20835,
 -108.20869,
 -108.20855,
 -108.20852,
 -108.20849,
 -108.20839,
 -108.20848,
 -108.20854,
 -108.20857,
 -108.36757,
 -108.38904,
 -108.44606,
 -108.47535,
 -108.53011,
 -108.61731,
 -108.65709,
 -108.70766,
 -108.71907,
 -108.72702,
 -108.73503,
 -108.73904,
 -108.7506,
 -108.75906,
 -108.82206,
 -108.86103,
 -108.88692,
 -109.00061,
 -109.05004,
 -109.05004,
 -109.04911,
 -109.04905,
 -109.04811,
 -109.04764,
 -109.04762,
 -109.0474,
 -109.04691,
 -109.047,
 -109.04692,
 -109.04633,
 -109.04603,
 -109.04598,
 -109.04644,
 -109.04662,
 -109.04636,
 -109.04621,
 -109.04606,
 -109.04578,
 -109.04601,
 -109.04575,
 -109.04579,
 -109.04547,
 -109.04544,
 -109.04531,
 -109.04522,
 -109.04522,
 -109.04518,
 -108.8978,
 -108.7496,
 -108.52833,
 -108.20233,
 -107.8669,
 -107.49519,
 -107.4121,
 -107.25094,
 -107.00562,
 -106.89142,
 -106.73137,
 -106.61906,
 -106.4654,
 -106.15386,
 -105.93761,
 -105.6576,
 -105.50752,
 -105.15658,
 -104.76311,
 -104.49245,
 -104.17265,
 -103.92627,
 -103.53892,
 -103.26061,
 -103.01356,
 -103.0022,
 -103.0022,
 -103.00232,
 -103.00232,
 -103.00228,
 -103.00227,
 -103.00224,
 -103.00223,
 -103.00218,
 -103.00206,
 -103.00213,
 -103.00215,
 -103.0021,
 -103.00214,
 -103.00218,
 -103.00237,
 -103.00218,
 -103.0023,
 -103.00226,
 -103.00223,
 -103.00226,
 -103.0023,
 -103.00233,
 -103.00233,
 -103.00228,
 -103.00228,
 -103.00245,
 -103.00243,
 -103.02394,
 -103.04133,
 -103.04249,
 -103.04239,
 -103.04283,
 -103.04312,
 -103.04338,
 -103.04362,
 -103.04374,
 -103.04376,
 -103.04993,
 -103.05727,
 -103.06464,
 -103.06478,
 -103.53275]

ut_lons = [-114.04392,
 -114.04391,
 -114.04375,
 -114.04195,
 -114.04061,
 -114.04055,
 -114.0398,
 -114.04172,
 -114.0391,
 -113.80254,
 -113.64886,
 -113.49562,
 -113.36362,
 -113.20505,
 -113.10627,
 -112.96233,
 -112.83266,
 -112.78175,
 -112.68558,
 -112.58229,
 -112.45023,
 -112.26534,
 -112.1985,
 -112.10309,
 -111.98965,
 -111.9304,
 -111.88098,
 -111.82932,
 -111.73177,
 -111.51913,
 -111.4087,
 -111.26009,
 -111.14884,
 -111.04934,
 -111.04669,
 -111.04682,
 -111.04631,
 -111.04601,
 -111.0458,
 -111.04611,
 -111.04648,
 -111.04667,
 -111.04688,
 -111.04686,
 -110.94406,
 -110.86384,
 -110.70521,
 -110.55878,
 -110.43401,
 -110.34177,
 -110.25071,
 -110.14713,
 -110.0004,
 -109.90645,
 -109.75044,
 -109.67348,
 -109.63381,
 -109.51776,
 -109.43099,
 -109.30329,
 -109.05847,
 -109.05008,
 -109.04907,
 -109.04899,
 -109.05051,
 -109.05088,
 -109.05093,
 -109.05088,
 -109.05087,
 -109.05077,
 -109.05132,
 -109.05077,
 -109.05119,
 -109.05158,
 -109.05305,
 -109.05655,
 -109.06015,
 -109.06017,
 -109.04984,
 -109.0498,
 -109.04232,
 -109.04304,
 -109.04159,
 -109.04191,
 -109.04582,
 -109.04583,
 -109.04519,
 -109.04531,
 -109.04522,
 -109.04538,
 -109.26993,
 -109.43568,
 -109.83491,
 -110.13851,
 -110.272,
 -110.54945,
 -110.73783,
 -111.03957,
 -111.2523,
 -111.39598,
 -111.58602,
 -111.99142,
 -112.32985,
 -112.48455,
 -112.75086,
 -112.96895,
 -112.99281,
 -113.02079,
 -113.17698,
 -113.32097,
 -113.4727,
 -113.62465,
 -113.86852,
 -113.94557,
 -114.05052,
 -114.0506,
 -114.05187,
 -114.05264,
 -114.05198,
 -114.04939,
 -114.05013,
 -114.04997,
 -114.04992,
 -114.04916,
 -114.04833,
 -114.04885,
 -114.04841,
 -114.04779,
 -114.0473,
 -114.04757,
 -114.04727,
 -114.04658,
 -114.04644,
 -114.04619,
 -114.04558]

##### Latitude Data

In [18]:
az_lats = [34.87057,
 35.00186,
 35.00332,
 35.07971,
 35.11791,
 35.14231,
 35.21811,
 35.37012,
 35.49125,
 35.5417,
 35.60517,
 35.66942,
 35.73579,
 35.81412,
 35.86436,
 35.9167,
 35.98542,
 36.05493,
 36.11546,
 36.15058,
 36.10119,
 36.01461,
 36.09833,
 36.21464,
 36.37619,
 36.60322,
 36.817,
 36.99997,
 37.0004,
 37.0004,
 36.99998,
 36.99998,
 36.99998,
 36.99998,
 36.99998,
 36.99998,
 37.00022,
 37.00017,
 37.00012,
 37.00048,
 37.00094,
 37.00105,
 37.00097,
 37.00166,
 37.00147,
 37.00102,
 37.00247,
 37.00325,
 37.00383,
 36.99828,
 36.99845,
 36.99831,
 36.9991,
 36.99926,
 36.99908,
 36.99908,
 36.99897,
 36.8531,
 36.70384,
 36.54513,
 36.41637,
 36.29154,
 36.18724,
 36.03128,
 35.93088,
 35.81044,
 35.65092,
 35.45859,
 35.30697,
 34.91388,
 34.71264,
 34.44558,
 34.08446,
 33.71335,
 33.3477,
 33.07165,
 32.70386,
 32.40743,
 32.1771,
 31.87069,
 31.63698,
 31.3325,
 31.33252,
 31.3338,
 31.33396,
 31.334,
 31.33394,
 31.33406,
 31.33393,
 31.33408,
 31.33399,
 31.33341,
 31.33363,
 31.33296,
 31.33299,
 31.33305,
 31.33363,
 31.33328,
 31.33281,
 31.33283,
 31.33257,
 31.34898,
 31.38586,
 31.40231,
 31.41305,
 31.42333,
 31.43196,
 31.45068,
 31.46195,
 31.4678,
 31.47995,
 31.49099,
 31.50825,
 31.51945,
 31.54305,
 31.56227,
 31.6012,
 31.61823,
 31.62425,
 31.63623,
 31.65645,
 31.67094,
 31.67701,
 31.69377,
 31.72891,
 31.75165,
 31.76301,
 31.78954,
 31.80473,
 31.82357,
 31.83702,
 31.8513,
 31.86132,
 31.87666,
 31.88514,
 31.89671,
 31.90787,
 31.99917,
 32.0054,
 32.02905,
 32.12566,
 32.17992,
 32.21797,
 32.2376,
 32.28088,
 32.31044,
 32.32538,
 32.33509,
 32.36468,
 32.43408,
 32.48373,
 32.49526,
 32.50602,
 32.50999,
 32.51839,
 32.52419,
 32.53277,
 32.54351,
 32.55396,
 32.56772,
 32.56625,
 32.56133,
 32.57093,
 32.58137,
 32.5955,
 32.60317,
 32.61295,
 32.6238,
 32.62325,
 32.6247,
 32.63705,
 32.65006,
 32.66489,
 32.66985,
 32.67414,
 32.6785,
 32.68221,
 32.68517,
 32.68732,
 32.6986,
 32.70545,
 32.71192,
 32.71829,
 32.71943,
 32.73946,
 32.73137,
 32.73487,
 32.87408,
 32.97206,
 33.03255,
 33.23376,
 33.39691,
 33.47131,
 33.58709,
 33.6969,
 33.84446,
 33.91285,
 33.96372,
 34.04257,
 34.12866,
 34.17212,
 34.31087,
 34.41527,
 34.47903,
 34.64288,
 34.71453,
 34.79181,
 34.86997]

co_lats = [38.215,
 38.40118,
 38.60929,
 38.81393,
 38.95788,
 39.11656,
 39.22605,
 39.36423,
 39.56752,
 39.79876,
 40.03782,
 40.18844,
 40.2929,
 40.41493,
 40.50615,
 40.68445,
 40.87296,
 41.00066,
 41.00051,
 41.00013,
 41.0001,
 40.99996,
 41.00008,
 41.00011,
 41.00139,
 41.00205,
 41.00197,
 41.00228,
 41.00305,
 41.00283,
 41.00305,
 41.00315,
 41.00213,
 40.99927,
 40.997,
 40.99701,
 40.99722,
 40.99766,
 40.99818,
 40.9983,
 40.99826,
 41.00153,
 41.00162,
 41.00139,
 41.00141,
 41.00153,
 41.00166,
 41.0017,
 41.00185,
 41.00191,
 41.00239,
 41.00231,
 41.00234,
 41.00243,
 41.00245,
 41.00235,
 41.00247,
 40.97899,
 40.9255,
 40.87501,
 40.82214,
 40.75901,
 40.71742,
 40.69984,
 40.65115,
 40.61334,
 40.55651,
 40.48846,
 40.31022,
 40.1605,
 40.01399,
 40.00308,
 40.00061,
 39.87098,
 39.67645,
 39.53889,
 39.40853,
 39.33572,
 39.24305,
 38.84279,
 38.66387,
 38.59586,
 38.52933,
 38.47219,
 38.41997,
 38.30634,
 38.26504,
 38.17556,
 38.05734,
 37.94303,
 37.80941,
 37.68749,
 37.49347,
 37.45597,
 37.3385,
 37.22191,
 37.10676,
 36.99352,
 36.99302,
 36.99306,
 36.99305,
 36.99311,
 36.99325,
 36.99357,
 36.99367,
 36.9942,
 36.99439,
 36.99425,
 36.99441,
 36.99457,
 36.99479,
 36.99494,
 36.99487,
 36.99513,
 36.99504,
 36.9952,
 36.99557,
 36.99885,
 36.99977,
 36.99959,
 36.99944,
 36.99908,
 36.99855,
 36.99852,
 37.0001,
 37.0001,
 37.00021,
 36.99956,
 36.99892,
 36.99669,
 36.99511,
 36.99367,
 36.9935,
 36.99527,
 36.99589,
 36.99578,
 36.99556,
 36.99469,
 36.99378,
 36.99303,
 36.99263,
 37.00014,
 37.00001,
 37.00001,
 37.0,
 37.00001,
 37.0,
 36.99924,
 36.99927,
 36.99887,
 36.99885,
 36.99908,
 36.99908,
 37.01706,
 37.09597,
 37.20443,
 37.35363,
 37.5667,
 37.7899,
 37.97469,
 38.10272,
 38.21472]
nm_lats = [32.00032,
 32.00001,
 32.00041,
 32.0005,
 32.00198,
 32.00076,
 31.98981,
 31.93601,
 31.90997,
 31.84661,
 31.8054,
 31.78318,
 31.78328,
 31.78391,
 31.78394,
 31.78395,
 31.78399,
 31.784,
 31.78409,
 31.78395,
 31.78393,
 31.78371,
 31.78394,
 31.78404,
 31.78385,
 31.78384,
 31.78379,
 31.78378,
 31.78381,
 31.7837,
 31.78369,
 31.78367,
 31.78367,
 31.78355,
 31.78354,
 31.78377,
 31.78367,
 31.78366,
 31.78367,
 31.78365,
 31.78367,
 31.78365,
 31.7836,
 31.78365,
 31.78366,
 31.78365,
 31.78369,
 31.78365,
 31.78374,
 31.78374,
 31.78374,
 31.78374,
 31.7837,
 31.78371,
 31.78371,
 31.78368,
 31.78368,
 31.78368,
 31.78361,
 31.78363,
 31.78365,
 31.7836,
 31.78359,
 31.78359,
 31.7836,
 31.78363,
 31.78365,
 31.78361,
 31.78357,
 31.78351,
 31.7835,
 31.78353,
 31.78357,
 31.78358,
 31.78359,
 31.7836,
 31.74465,
 31.72969,
 31.72323,
 31.7194,
 31.71733,
 31.68878,
 31.62561,
 31.58935,
 31.5608,
 31.52581,
 31.50105,
 31.4998,
 31.47883,
 31.42917,
 31.37656,
 31.3561,
 31.3334,
 31.3334,
 31.33341,
 31.33347,
 31.33344,
 31.33336,
 31.33329,
 31.33323,
 31.33319,
 31.33307,
 31.33298,
 31.3329,
 31.33285,
 31.33273,
 31.33261,
 31.33226,
 31.33232,
 31.33231,
 31.33217,
 31.33224,
 31.3325,
 31.63698,
 31.87069,
 32.1771,
 32.40743,
 32.70386,
 33.07165,
 33.3477,
 33.71335,
 34.08446,
 34.44558,
 34.71264,
 34.91388,
 35.30697,
 35.45859,
 35.65092,
 35.81044,
 35.93088,
 36.03128,
 36.18724,
 36.29154,
 36.41637,
 36.54513,
 36.70384,
 36.8531,
 36.99897,
 36.99908,
 36.99908,
 36.99885,
 36.99887,
 36.99927,
 36.99924,
 37.0,
 37.00001,
 37.0,
 37.00001,
 37.00001,
 37.00014,
 36.99263,
 36.99303,
 36.99378,
 36.99469,
 36.99556,
 36.99578,
 36.99589,
 36.99527,
 36.9935,
 36.99367,
 36.99511,
 36.99669,
 36.99892,
 36.99956,
 37.00021,
 37.0001,
 37.00006,
 36.9686,
 36.95031,
 36.91582,
 36.91402,
 36.90243,
 36.89755,
 36.86996,
 36.81735,
 36.76898,
 36.74295,
 36.7193,
 36.7013,
 36.68525,
 36.67653,
 36.65329,
 36.6278,
 36.60265,
 36.5919,
 36.56927,
 36.56386,
 36.54875,
 36.52963,
 36.51475,
 36.50609,
 36.50046,
 36.5004,
 36.50042,
 35.76515,
 35.21202,
 35.1362,
 34.88888,
 34.67259,
 34.53564,
 34.40999,
 34.27181,
 34.03983,
 33.71754,
 33.35051,
 33.00011,
 32.59516,
 32.00034]

ut_lats = [40.68928,
 40.68985,
 40.76026,
 41.05548,
 41.36,
 41.59062,
 41.89425,
 41.99372,
 41.99367,
 41.98895,
 41.99102,
 41.99331,
 41.99384,
 41.99645,
 41.99735,
 41.99841,
 41.99938,
 41.99973,
 42.00021,
 42.00054,
 42.00099,
 42.00111,
 42.00116,
 41.99763,
 41.99834,
 41.99861,
 41.99856,
 41.99875,
 41.99926,
 41.99951,
 42.00063,
 42.00132,
 42.00154,
 42.00159,
 42.00157,
 42.00034,
 41.83664,
 41.64119,
 41.52149,
 41.41513,
 41.36572,
 41.20441,
 41.10318,
 41.00917,
 40.9976,
 40.99725,
 40.99635,
 40.99635,
 40.99485,
 40.99533,
 40.99609,
 40.99634,
 40.99734,
 40.99766,
 40.99797,
 40.99841,
 40.9983,
 40.99871,
 40.99964,
 41.00064,
 41.00069,
 41.00066,
 40.87296,
 40.68445,
 40.50615,
 40.41493,
 40.2929,
 40.18844,
 40.03782,
 39.79876,
 39.56752,
 39.36423,
 39.22605,
 39.11656,
 38.95788,
 38.81393,
 38.60929,
 38.40118,
 38.215,
 38.21472,
 38.10272,
 37.97469,
 37.7899,
 37.5667,
 37.35363,
 37.20443,
 37.09597,
 37.01706,
 36.99908,
 36.99908,
 36.99926,
 36.9991,
 36.99831,
 36.99845,
 36.99828,
 37.00383,
 37.00325,
 37.00247,
 37.00102,
 37.00147,
 37.00166,
 37.00097,
 37.00105,
 37.00094,
 37.00048,
 37.00012,
 37.00017,
 37.00022,
 36.99998,
 36.99998,
 36.99998,
 36.99998,
 36.99998,
 36.99998,
 37.0004,
 37.0004,
 37.13439,
 37.47222,
 37.70735,
 37.77873,
 37.95499,
 38.20495,
 38.55049,
 38.75165,
 38.90545,
 39.08777,
 39.23851,
 39.36296,
 39.45715,
 39.61018,
 39.75817,
 39.99994,
 40.09896,
 40.30302,
 40.4958]

In [19]:
# Create the figure: p
p = figure(x_axis_label='latitude (degrees)', y_axis_label='longitude (degrees)')

# Create a list of az_lons, co_lons, nm_lons and ut_lons: x
x = [az_lons, co_lons, nm_lons, ut_lons]

# Create a list of az_lats, co_lats, nm_lats and ut_lats: y
y = [az_lats, co_lats, nm_lats, ut_lats]

# Add patches to figure p with line_color=white for x and y
p.patches(x,y,line_color='white')

# Specify the name of the output file and show the result
output_file('four_corners.html')
show(p)

### Data Formats

#### Plotting data from NumPy arrays
In the previous exercises, you made plots using data stored in lists. You learned that Bokeh can plot both numbers and datetime objects.

In this exercise, you'll generate NumPy arrays using np.linspace() and np.cos() and plot them using the circle glyph.

np.linspace() is a function that returns an array of evenly spaced numbers over a specified interval. For example, np.linspace(0, 10, 5) returns an array of 5 evenly spaced samples calculated over the interval [0, 10]. np.cos(x) calculates the element-wise cosine of some array x.

For more information on NumPy functions, you can refer to the NumPy User Guide and NumPy Reference.

The figure p has been provided for you.

In [20]:
# create figure
p = figure( x_axis_label='x', y_axis_label='y')

# Create array using np.linspace: x
x = np.linspace(0,5,100)

# Create array using np.cos: y
y = np.cos(x)

# Add circles at x and y
p.circle(x,y)

# Specify the name of the output file and show the result
output_file('./output/numpy.html')
show(p)

#### Plotting data from Pandas DataFrames
You can create Bokeh plots from Pandas DataFrames by passing column selections to the glyph functions.

Bokeh can plot floating point numbers, integers, and datetime data types. In this example, you will read a CSV file containing information on 392 automobiles manufactured in the US, Europe and Asia from 1970 to 1982.

The CSV file is provided for you as 'auto.csv'.

Your job is to plot miles-per-gallon (mpg) vs horsepower (hp) by passing Pandas column selections into the p.circle() function. Additionally, each glyph will be colored according to values in the color column.

In [21]:
# Create the figure: p
p = figure(x_axis_label='HP', y_axis_label='MPG')

# Plot mpg vs hp by color
p.circle(auto['hp'],auto['mpg'],size=10,color=auto['color'])

# Specify the name of the output file and show the result
output_file('./output/auto-df.html')
show(p)

#### The Bokeh ColumnDataSource (continued)
You can create a ColumnDataSource object directly from a Pandas DataFrame by passing the DataFrame to the class initializer.

In this exercise, we have imported pandas as pd and read in a data set containing all Olympic medals awarded in the 100 meter sprint from 1896 to 2012. A color column has been added indicating the CSS colorname we wish to use in the plot for every data point.

Your job is to import the ColumnDataSource class, create a new ColumnDataSource object from the DataFrame df, and plot circle glyphs with 'Year' on the x-axis and 'Time' on the y-axis. Color each glyph by the color column.

The figure object p has already been created for you.

In [22]:
# Import the ColumnDataSource class from bokeh.plotting
from bokeh.plotting import ColumnDataSource

# Create a ColumnDataSource from df: source
source = ColumnDataSource(olympic)

#create figure p
p = figure(x_axis_label='Year',y_axis_label='Time')

# Add circle glyphs to the figure p
p.circle('Year','Time',size=8,color='color',source=source)

# Specify the name of the output file and show the result
output_file('./output/sprint.html')
show(p)

### Customizing glyphs
#### Selection and non-selection glyphs
In this exercise, you're going to add the box_select tool to a figure and change the selected and non-selected circle glyph properties so that selected glyphs are red and non-selected glyphs are transparent blue.

You'll use the ColumnDataSource object of the Olympic Sprint dataset you made in the last exercise. It is provided to you with the name source.

After you have created the figure, be sure to experiment with the Box Select tool you added! As in previous exercises, you may have to scroll down to view the lower portion of the figure.

In [23]:
# Create a figure with the "box_select" tool: p
p = figure(x_axis_label='Year',y_axis_label='Time',tools='box_select')

# Add circle glyphs to the figure p with the selected and non-selected properties
p.circle('Year','Time',source=source,selection_color='red',
            nonselection_color='blue',nonselection_alpha=.1)

# Specify the name of the output file and show the result
output_file('./output/selection_glyph.html')
show(p)


#### Hover glyphs
Now let's practice using and customizing the hover tool.

In this exercise, you're going to plot the blood glucose levels for an unknown patient. The blood glucose levels were recorded every 5 minutes on October 7th starting at 3 minutes past midnight.

The date and time of each measurement are provided to you as x and the blood glucose levels in mg/dL are provided as y.

A bokeh figure is also provided in the workspace as p.

Your job is to add a circle glyph that will appear red when the mouse is hovered near the data points. You will also add a customized hover tool object to the plot.

When you're done, play around with the hover tool you just created! Notice how the points where your mouse hovers over turn red.

In [24]:
glucose['datetime'] =glucose['datetime'].astype('datetime64[ns]')
x = glucose['datetime'].values
x = np.arange(len(x))
y=glucose['glucose'].values

In [25]:
# import the HoverTool
from bokeh.models import HoverTool

p = figure(x_axis_label='Time of day (ms)',y_axis_label='Blood glucose (mg/dL)')

# Add circle glyphs to figure p
p.circle(x, y, size=10,
         fill_color='grey', alpha=.1, line_color=None,
         hover_fill_color='firebrick', hover_alpha=.5,
         hover_line_color='white')

# Create a HoverTool: hover
hover = HoverTool(tooltips=None,mode='vline')

# Add the hover tool to the figure p
p.add_tools(hover)

# Specify the name of the output file and show the result
output_file('./output/hover_glyph.html')
show(p)

#### Colormapping
The final glyph customization we'll practice is using the CategoricalColorMapper to color each glyph by a categorical property.

Here, you're going to use the automobile dataset to plot miles-per-gallon vs weight and color each circle glyph by the region where the automobile was manufactured.

The origin column will be used in the ColorMapper to color automobiles manufactured in the US as blue, Europe as red and Asia as green.

The automobile data set is provided to you as a Pandas DataFrame called df. The figure is provided for you as p.

In [26]:
#Import CategoricalColorMapper from bokeh.models
from bokeh.models import CategoricalColorMapper

# Convert df to a ColumnDataSource: source
source = ColumnDataSource(auto)

p = figure(x_axis_label='Weight',y_axis_label='Mpg')

# Make a CategoricalColorMapper object: color_mapper
color_mapper = CategoricalColorMapper(factors=['Europe', 'Asia', 'US'],
                                      palette=['red', 'green', 'blue'])

# Add a circle glyph to the figure p
p.circle('weight', 'mpg', source=source,
            color=dict(field='origin', transform=color_mapper),
            legend='origin')

# Specify the name of the output file and show the result
output_file('./output/colormap.html')
show(p)

## Chapter 2:Layouts, Interactions, and Annotations

### Introduction to layouts
Learn how to combine mutiple Bokeh plots into different kinds of layouts on a page, how to easily link different plots together in various ways, and how to add annotations such as legends and hover tooltips.

#### Creating rows of plots
Layouts are collections of Bokeh figure objects.

In this exercise, you're going to create two plots from the Literacy and Birth Rate data set to plot fertility vs female literacy and population vs female literacy.

By using the row() method, you'll create a single layout of the two figures.

Remember, as in the previous chapter, once you have created your figures, you can interact with them in various ways.

In this exercise, you may have to scroll sideways to view both figures in the row layout. Alternatively, you can view the figures in a new window by clicking on the expand icon to the right of the "Bokeh plot" tab.

In [27]:
fem_lit.head()

Country  Continent female literacy fertility    population
0      Chine       ASI            90.5     1.769  1.324655e+09
1       Inde       ASI            50.8     2.682  1.139965e+09
2        USA       NAM              99     2.077  3.040600e+08
3  Indonésie       ASI            88.8     2.132  2.273451e+08
4     Brésil       LAT            90.2     1.827  1.919715e+08

In [55]:
fem_lit1 = fem_lit.dropna()
fem_lit1.fertility = fem_lit1.fertility.astype('float64')
fem_lit1['female literacy'] = fem_lit1['female literacy'].astype('float64')
fem_lit1 = fem_lit1.rename(index=str, columns={"female literacy": "female_literacy", "Country ": "Country"})
fem_lit1.columns

/usr1/anaconda3/envs/jupyter_env/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


Index(['Country', 'Continent', 'female_literacy', 'fertility', 'population'], dtype='object')

In [44]:
# Import row from bokeh.layouts
from bokeh.layouts import row


# Convert df to a ColumnDataSource: source
source = ColumnDataSource(fem_lit1)

# Create the first figure: p1
p1 = figure(x_axis_label='fertility (children per woman)', y_axis_label='female_literacy (% population)')

# Add a circle glyph to p1
p1.circle('fertility', 'female_literacy', source=source)

# Create the second figure: p2
p2 = figure(x_axis_label='population', y_axis_label='female_literacy (% population)')

# Add a circle glyph to p2
p2.circle('population', 'female_literacy', source=source)

# Put p1 and p2 into a horizontal row: layout
layout = row(p1,p2)

# Specify the name of the output_file and show the result
output_file('./output/fert_row.html')
show(layout)

#### Creating columns of plots
In this exercise, you're going to use the column() function to create a single column layout of the two plots you created in the previous exercise.

Figure p1 has been created for you.

In this exercise and the ones to follow, you may have to scroll down to view the lower portion of the figure.

In [45]:
# Import column from the bokeh.layouts module
from bokeh.layouts import column

# Convert df to a ColumnDataSource: source
#source = ColumnDataSource(fem_lit1)

# Create a blank figure: p1
p1 = figure(x_axis_label='fertility (children per woman)', y_axis_label='female_literacy (% population)')

# Add circle scatter to the figure p1
p1.circle('fertility', 'female_literacy', source=source)

# Create a new blank figure: p2
p2 = figure(x_axis_label='population', y_axis_label='female_literacy (% population)')

# Add circle scatter to the figure p2
p2.circle('population', 'female_literacy', source=source)

# Put plots p1 and p2 in a column: layout
layout = column(p1,p2)

# Specify the name of the output_file and show the result
output_file('./output/fert_column.html')
show(layout)

#### Nesting rows and columns of plots
You can create nested layouts of plots by combining row and column layouts.

In this exercise, you'll make a 3-plot layout in two rows using the auto-mpg data set.

Three plots have been created for you of average mpg vs year, mpg vs hp, and mpg vs weight.

Your job is to use the column() and row() functions to make a two-row layout where the first row will have only the average mpg vs year plot and the second row will have mpg vs hp and mpg vs weight plots as columns.

By using the sizing_mode argument, you can scale the widths to fill the whole figure.

In [46]:
# Convert df to a ColumnDataSource: source
source = ColumnDataSource(auto)

####################
# Create the figure: p
mpg_hp = figure(x_axis_label='HP', y_axis_label='MPG')

# Plot mpg vs hp by color
mpg_hp.circle(auto['hp'],auto['mpg'],size=10,color=auto['color'])

####################
mpg_weight = figure(x_axis_label='Weight',y_axis_label='Mpg')

# Make a CategoricalColorMapper object: color_mapper
color_mapper = CategoricalColorMapper(factors=['Europe', 'Asia', 'US'],
                                      palette=['red', 'green', 'blue'])

# Add a circle glyph to the figure p
mpg_weight.circle('weight', 'mpg', source=source,
            color=dict(field='origin', transform=color_mapper),
            legend='origin')

#####################
df_auto = auto.groupby(['yr']).mean().reset_index()
x = df_auto['yr'].values
y = df_auto['mpg']
# Create a figure with x_axis_type='datetime': p
avg_mpg = figure(x_axis_label='year', y_axis_label='Avg_mpg')

# Plot date along the x-axis and price along the y-axis
avg_mpg.line(x,y)
#####################

# Make a column layout that will be used as the second row: row2
row2 = column([mpg_hp, mpg_weight], sizing_mode='scale_width')

# Make a row layout that includes the above column layout: layout
layout = row([avg_mpg, row2], sizing_mode='scale_width')

# Specify the name of the output_file and show the result
output_file('./output/layout_custom.html')
show(layout)

### Advanced Layouts

#### Creating gridded layouts
Regular grids of Bokeh plots can be generated with gridplot.

In this example, you're going to display four plots of fertility vs female literacy for four regions: Latin America, Africa, Asia and Europe.

Your job is to create a list-of-lists for the four Bokeh plots that have been provided to you as p1, p2, p3 and p4. The list-of-lists defines the row and column placement of each plot.

In [50]:
# Import gridplot from bokeh.layouts
from bokeh.layouts import gridplot

fertility_asia = df[df['Continent'] == 'ASI']['fertility'].values
female_literacy_asia = df[df['Continent'] == 'ASI']['female literacy'].values
fertility_eu = df[df['Continent'] == 'EUR']['fertility'].values
female_literacy_eu = df[df['Continent'] == 'EUR']['female literacy'].values
#################
# Create the figure: p
p1 = figure(title="Latin America",x_axis_label='fertility (children per woman)', 
            y_axis_label='female_literacy (% population)',plot_width=300, plot_height=300)

# Add a blue circle glyph to the figure p
p1.circle(fertility_latinamerica, female_literacy_latinamerica, color='blue', alpha=.8, size=3)

####################
# Create the figure: p
p2 = figure(title="Africa",x_axis_label='fertility (children per woman)', 
            y_axis_label='female_literacy (% population)',plot_width=300, plot_height=300)

# Add a red circle glyph to the figure p
p2.circle(fertility_africa, female_literacy_africa, color='blue', alpha=.8, size=3)
###########################
# Create the figure: p
p3 = figure(title="Asia",x_axis_label='fertility (children per woman)', 
            y_axis_label='female_literacy (% population)',plot_width=300, plot_height=300)

# Add a blue circle glyph to the figure p
p3.circle(fertility_asia, female_literacy_asia, color='blue', alpha=.8, size=3)

#######################
# Create the figure: p
p4 = figure(title="Europe",x_axis_label='fertility (children per woman)', 
            y_axis_label='female_literacy (% population)',plot_width=300, plot_height=300)

# Add a blue circle glyph to the figure p
p4.circle(fertility_eu, female_literacy_eu, color='blue', alpha=.8, size=3)

#################


# Create a list containing plots p1 and p2: row1
row1 = [p1,p2]

# Create a list containing plots p3 and p4: row2
row2 = [p3,p4]

# Create a gridplot using row1 and row2: layout
layout = gridplot([row1,row2])

# Specify the name of the output_file and show the result
output_file('./output/grid.html')
show(layout)

#### Starting tabbed layouts
Tabbed layouts can be created in Bokeh by placing plots or layouts in Panels.

In this exercise, you'll take the four fertility vs female literacy plots from the last exercise and make a Panel() for each.

No figure will be generated in this exercise. Instead, you will use these panels in the next exercise to build and display a tabbed layout.

In [51]:
# Import Panel from bokeh.models.widgets
from bokeh.models.widgets import Panel

# Create tab1 from plot p1: tab1
tab1 = Panel(child=p1, title='Latin America')

# Create tab2 from plot p2: tab2
tab2 = Panel(child=p2, title='Africa')

# Create tab3 from plot p3: tab3
tab3 = Panel(child=p3, title='Asia')

# Create tab4 from plot p4: tab4
tab4 = Panel(child=p4, title='Europe')

#### Displaying tabbed layouts
Tabbed layouts are collections of Panel objects. Using the figures and Panels from the previous two exercises, you'll create a tabbed layout to change the region in the fertility vs female literacy plots.

Your job is to create the layout using Tabs() and assign the tabs keyword argument to your list of Panels. The Panels have been created for you as tab1, tab2, tab3 and tab4.

After you've displayed the figure, explore the tabs you just added! The "Pan", "Box Zoom" and "Wheel Zoom" tools are also all available as before.

In [52]:
# Import Tabs from bokeh.models.widgets
from bokeh.models.widgets import Tabs

# Create a Tabs layout: layout
layout = Tabs(tabs=[tab1, tab2, tab3, tab4])

# Specify the name of the output_file and show the result
output_file('./output/tabs.html')
show(layout)

### Linking plots together

#### Linked axes
Linking axes between plots is achieved by sharing range objects.

In this exercise, you'll link four plots of female literacy vs fertility so that when one plot is zoomed or dragged, one or more of the other plots will respond.

The four plots p1, p2, p3 and p4 along with the layout that you created in the last section have been provided for you.

Your job is link p1 with the three other plots by assignment of the .x_range and .y_range attributes.

After you have linked the axes, explore the plots by clicking and dragging along the x or y axes of any of the plots, and notice how the linked plots change together.


In [34]:
# Create a list containing plots p1 and p2: row1
row1 = [p1,p2]

# Create a list containing plots p3 and p4: row2
row2 = [p3,p4]

# Create a gridplot using row1 and row2: layout
layout = gridplot([row1,row2])

# Link the x_range of p2 to p1: p2.x_range
p2.x_range = p1.x_range

# Link the y_range of p2 to p1: p2.y_range
p2.y_range = p1.y_range

# Link the x_range of p3 to p1: p3.x_range
p3.x_range = p1.x_range

# Link the y_range of p4 to p1: p4.y_range
p4.y_range = p1.y_range

# Specify the name of the output_file and show the result
output_file('./output/linked_range.html')
show(layout)

#### Linked brushing
By sharing the same ColumnDataSource object between multiple plots, selection tools like BoxSelect and LassoSelect will highlight points in both plots that share a row in the ColumnDataSource.

In this exercise, you'll plot female literacy vs fertility and population vs fertility in two plots using the same ColumnDataSource.

After you have built the figure, experiment with the Lasso Select and Box Select tools. Use your mouse to drag a box or lasso around points in one figure, and notice how points in the other figure that share a row in the ColumnDataSource also get highlighted.

Before experimenting with the Lasso Select, however, click the Bokeh plot pop-out icon to pop out the figure so that you can definitely see everything that you're doing.

In [35]:
# Create ColumnDataSource: source
source = ColumnDataSource(fem_lit1)

# Create the first figure: p1
p1 = figure(x_axis_label='fertility (children per woman)', y_axis_label='female literacy (% population)',
            tools='box_select,lasso_select',plot_width=300, plot_height=300)

# Add a circle glyph to p1
p1.circle('fertility','female literacy',source=source)

# Create the second figure: p2
p2 = figure(x_axis_label='fertility (children per woman)', y_axis_label='population (millions)',
            tools='box_select,lasso_select',plot_width=300, plot_height=300)

# Add a circle glyph to p2
p2.circle('fertility','population',source=source)

# Create row layout of figures p1 and p2: layout
layout = row(p1,p2)

# Specify the name of the output_file and show the result
output_file('./output/linked_brush.html')
show(layout)

### Annotations and guides
#### How to create legends
Legends can be added to any glyph by using the legend keyword argument.

In this exercise, you will plot two circle glyphs for female literacy vs fertility in Africa and Latin America.

Two ColumnDataSources called latin_america and africa have been provided.

Your job is to plot two circle glyphs for these two objects with fertility on the x axis and female_literacy on the y axis and add the legend values. The figure p has been provided for you.

In [56]:
# Create ColumnDataSource: source
#fem_lit1.columns = ['Country', 'Continent', 'female literacy', 'fertility', 'population']

lat = fem_lit1[fem_lit1['Continent'] == 'LAT']
af = fem_lit1[fem_lit1['Continent'] == 'AF']

'''af["Country1"] =  [s.encode('ascii', 'ignore').strip()
               for s in af['Country'].str.decode('unicode_escape')]
lat["Country1"] =  [s.encode('ascii', 'ignore').strip()
               for s in lat['Country'].str.decode('unicode_escape')]'''

latin_america = ColumnDataSource(lat)
africa = ColumnDataSource(af)

p = figure(x_axis_label='fertility (children per woman)', 
            y_axis_label='female_literacy (% population)',plot_width=500, plot_height=500)

# Add the first circle glyph to the figure p
p.circle('fertility', 'female_literacy', source=latin_america, size=5, color='red', legend='Latin America')

# Add the second circle glyph to the figure p
p.circle('fertility', 'female_literacy', source=africa, size=5, color='blue', legend='Africa')

# Specify the name of the output_file and show the result
output_file('./output/fert_lit_groups.html')
show(p)

In [37]:
af.columns

Index(['Country', 'Continent', 'female literacy', 'fertility', 'population'], dtype='object')

#### Positioning and styling legends
Properties of the legend can be changed by using the legend member attribute of a Bokeh figure after the glyphs have been plotted.

In this exercise, you'll adjust the background color and legend location of the female literacy vs fertility plot from the previous exercise.

The figure object p has been created for you along with the circle glyphs.

In [57]:
# Assign the legend to the bottom left: p.legend.location
p.legend.location = 'bottom_left'

# Fill the legend background with the color 'lightgray': p.legend.background_fill_color
p.legend.background_fill_color = 'lightgray'

# Specify the name of the output_file and show the result
output_file('./output/fert_lit_groups.html')
show(p)

#### Adding a hover tooltip
Working with the HoverTool is easy for data stored in a ColumnDataSource.

In this exercise, you will create a HoverTool object and display the country for each circle glyph in the figure that you created in the last exercise. This is done by assigning the tooltips keyword argument to a list-of-tuples specifying the label and the column of values from the ColumnDataSource using the @ operator.

The figure object has been prepared for you as p.

After you have added the hover tooltip to the figure, be sure to interact with it by hovering your mouse over each point to see which country it represents.

In [59]:
# Create a HoverTool object: hover
hover = HoverTool(tooltips=[('Continent','@Continent'),('fertility','@fertility'),('population','@population'),
                           ('Country','@Country'),('female literacy','@female_literacy')])

# Add the HoverTool object to figure p
p.add_tools(hover)

# Specify the name of the output_file and show the result
output_file('./output/hover.html')
show(p)

## Chapter 3:Building interactive apps with Bokeh

### Introducing the Bokeh Server

#### Using the current document
Let's get started with building an interactive Bokeh app. This typically begins with importing the curdoc, or "current document", function from bokeh.io. This current document will eventually hold all the plots, controls, and layouts that you create. Your job in this exercise is to use this function to add a single plot to your application.

In the video, Bryan described the process for running a Bokeh app using the bokeh serve command line tool. In this chapter and the one that follows, the DataCamp environment does this for you behind the scenes. Notice that your code is part of a script.py file. When you hit 'Submit Answer', you'll see in the IPython Shell that we call ```bokeh serve script.py``` for you.

Remember, as in the previous chapters, that there are different options available for you to interact with your plots, and as before, you may have to scroll down to view the lower portion of the plots.